source/d1cel.cpp

```diff
@@ -138,3 +140,231 @@ bool D1Cel::load(QString filePath, OpenAsParam *params)
+        clipped.append((clippedForced && params->clipped == SAVE_CLIPPING_TYPE::CLIPPED_TRUE) || (!clippedForced && frame->isClipped()));
+    }
+    // calculate header size
+    int HEADER_SIZE = 4 + 4 + numFrames * 4;
+    // calculate sub header size
+    int subHeaderSize = SUB_HEADER_SIZE;
+    for (int n = 0; n < numFrames; n++) {
+        QPointer<D1CelFrameBase> frame = this->frames[n];
+        if (clipped[n]) {
+            int hs = (frame->getHeight() - 1) / CEL_BLOCK_HEIGHT;
+            hs = (hs + 1) * sizeof(quint16);
+            subHeaderSize = std::max(subHeaderSize, hs);
+        }
+    }
+    // estimate data size
+    int maxSize = HEADER_SIZE;
+    for (int n = 0; n < numFrames; n++) {
+        QPointer<D1CelFrameBase> frame = this->frames[n];
+        if (clipped[n]) {
+            maxSize += subHeaderSize; // SUB_HEADER_SIZE
+        }
+        maxSize += frame->getHeight() * (2 * frame->getWidth());
+    }
+
+    QByteArray fileData;
+    fileData.append(maxSize, 0);
+
+    quint8 *buf = (quint8 *)fileData.data();
+    *(quint32 *)&buf[0] = SwapLE32(numFrames);
+    *(quint32 *)&buf[4] = SwapLE32(HEADER_SIZE);
+    quint8 *pBuf = &buf[HEADER_SIZE];
+    for (int n = 0; n < numFrames; n++) {
+        pBuf = writeFrameData((D1CelFrame *)this->getFrame(n), pBuf, subHeaderSize, clipped[n]);
+        *(quint32 *)&buf[4 + 4 * (n + 1)] = SwapLE32(pBuf - buf);
+    }
+
+    // write to file
+    QDataStream out(&outFile);
+    out.writeRawData((char *)buf, pBuf - buf);
+
+    return true;
+}
+
+bool D1Cel::writeCompFileData(QFile &outFile, SaveAsParam *params)
+{
+    const int numFrames = this->frames.count();
+
+    // prepare clipping info
+    bool clippedForced = params != nullptr && params->clipped != SAVE_CLIPPING_TYPE::CLIPPED_AUTODETECT;
+    QList<bool> clipped;
+    for (int n = 0; n < numFrames; n++) {
+        QPointer<D1CelFrameBase> frame = this->frames[n];
+        clipped.append((clippedForced && params->clipped == SAVE_CLIPPING_TYPE::CLIPPED_TRUE) || (!clippedForced && frame->isClipped()));
+    }
+
+    // calculate header size
+    int numGroups;
+    int headerSize = 0;
+    QList<int> groupSizes;
+    if (params == nullptr || params->groupNum == 0) {
+        numGroups = this->getGroupCount();
+        for (int i = 0; i < numGroups; i++) {
+            QPair<quint16, quint16> gfi = this->getGroupFrameIndices(i);
+            int ni = gfi.second - gfi.first + 1;
+            groupSizes.append(ni);
+            headerSize += 4 + 4 * (ni + 1);
+        }
+    } else {
+        numGroups = params->groupNum;
+        if (numFrames % numGroups != 0) {
+            QMessageBox::critical(nullptr, "Error", "Frames can not be split to equal groups.");
+            return false;
+        }
+        for (int i = 0; i < numGroups; i++) {
+            int ni = numFrames / numGroups;
+            groupSizes.append(ni);
+            headerSize += 4 + 4 * (ni + 1);
+        }
+    }
+
+    // if (numGroups > 1) {
```

Comment:
Is this because it should not have been a compilation if there is only 1 group?

Reply:
Right, similar to the case above (line 357) I'm not exactly sure how the grouping supposed to work.